File: VB/RangeDistributionExample/Form1.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraMap

Namespace RangeDistributionExample

    Partial Public Class Form1
        Inherits Form

        Private Const LinearDistributionIndex As Integer = 0
        Private Const LogarithmicDistributionIndex As Integer = 1
        Private Const ExponentialDistributionIndex As Integer = 2

        Private Const filePath As String = "..\..\Countries.shp"
        Private fileLayer As New VectorFileLayer()
        Private ReadOnly Property Colorizer() As ChoroplethColorizer
            Get
                Return TryCast(fileLayer.Colorizer, ChoroplethColorizer)
            End Get
        End Property

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
            mapControl1.Layers.Add(fileLayer)
            mapControl1.Legends.Add(CreateLegend())

            fileLayer.FileLoader = CreateLoader(filePath)
            fileLayer.Colorizer = CreateColorizer()

            comboRangeDistribution.SelectedIndex = 0
        End Sub

        Private Function CreateLoader(ByVal filePath As String) As MapFileLoaderBase
            Dim baseUri As New Uri(System.Reflection.Assembly.GetEntryAssembly().Location)
            Dim loader As New ShapefileLoader() With {.FileUri = New Uri(baseUri, filePath)}
            Return loader
        End Function

        Private Function CreateColorizer() As MapColorizer

            Dim colorizer_Renamed As New ChoroplethColorizer() With { _
                .ApproximateColors = True, .RangeDistribution = New LinearRangeDistribution(), .ValueProvider = New ShapeAttributeValueProvider() With {.AttributeName = "GDP_MD_EST"} _
            }

            colorizer_Renamed.ColorItems.Add(New ColorizerColorItem(Color.Blue))
            colorizer_Renamed.ColorItems.Add(New ColorizerColorItem(Color.Red))

            colorizer_Renamed.RangeStops.Add(0)
            colorizer_Renamed.RangeStops.Add(1000)
            colorizer_Renamed.RangeStops.Add(100000)
            colorizer_Renamed.RangeStops.Add(10000000)
            colorizer_Renamed.RangeStops.Add(1000000000)

            Return colorizer_Renamed
        End Function

        Private Function CreateLegend() As MapLegendBase
            Dim legend As New ColorScaleLegend() With {.Header = "GDP in Countries", .Layer = fileLayer}

            Return legend
        End Function

        Private Sub checkApproximateColors_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles checkApproximateColors.CheckedChanged
            Colorizer.ApproximateColors = checkApproximateColors.Checked
        End Sub

        Private Sub OnParametersChanged(ByVal sender As Object, ByVal e As EventArgs) Handles tbFactor.TextChanged, comboRangeDistribution.SelectedIndexChanged
            ' Set selected range distribution and factor if it is needed.
            Select Case comboRangeDistribution.SelectedIndex
                Case LogarithmicDistributionIndex
                    tbFactor.Enabled = True
                    Colorizer.RangeDistribution = New LogarithmicRangeDistribution() With {.Factor = GetFactor()}
                Case LinearDistributionIndex
                    tbFactor.Enabled = True
                    Colorizer.RangeDistribution = New ExponentialRangeDistribution() With {.Factor = GetFactor()}
                Case Else
                    Colorizer.RangeDistribution = New LinearRangeDistribution()
                    tbFactor.Enabled = False
            End Select
        End Sub

        Private Function GetFactor() As Double
            Return If(String.IsNullOrEmpty(tbFactor.Text), 0, Convert.ToDouble(tbFactor.Text))
        End Function
    End Class
End Namespace

File: CS/RangeDistributionExample/Form1.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraMap;

namespace RangeDistributionExample {

    public partial class Form1 : Form {
        const int LinearDistributionIndex = 0;
        const int LogarithmicDistributionIndex = 1;
        const int ExponentialDistributionIndex = 2;

        const string filePath = "..\\..\\Countries.shp";
        VectorFileLayer fileLayer = new VectorFileLayer();
        ChoroplethColorizer Colorizer {
            get { return fileLayer.Colorizer as ChoroplethColorizer; }
        }

        public Form1() {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e) {
            mapControl1.Layers.Add(fileLayer);
            mapControl1.Legends.Add(CreateLegend());

            fileLayer.FileLoader = CreateLoader(filePath);
            fileLayer.Colorizer = CreateColorizer();

            comboRangeDistribution.SelectedIndex = 0;
        }

        private MapFileLoaderBase CreateLoader(string filePath) {
            Uri baseUri = new Uri(System.Reflection.Assembly.GetEntryAssembly().Location);
            ShapefileLoader loader = new ShapefileLoader() {
                FileUri = new Uri(baseUri, filePath)
            };
            return loader;
        }

        private MapColorizer CreateColorizer() {
            ChoroplethColorizer colorizer = new ChoroplethColorizer() {
                ApproximateColors = true,
                RangeDistribution = new LinearRangeDistribution(),
                ValueProvider = new ShapeAttributeValueProvider() { AttributeName = "GDP_MD_EST" }
            };

            colorizer.ColorItems.Add(new ColorizerColorItem(Color.Blue));
            colorizer.ColorItems.Add(new ColorizerColorItem(Color.Red));

            colorizer.RangeStops.Add(0);
            colorizer.RangeStops.Add(1000);
            colorizer.RangeStops.Add(100000);
            colorizer.RangeStops.Add(10000000);
            colorizer.RangeStops.Add(1000000000);

            return colorizer;
        }

        private MapLegendBase CreateLegend() {
            ColorScaleLegend legend = new ColorScaleLegend() {
                Header = "GDP in Countries",
                Layer = fileLayer
            };

            return legend;
        }

        private void checkApproximateColors_CheckedChanged(object sender, EventArgs e) {
            Colorizer.ApproximateColors = checkApproximateColors.Checked;
        }

        private void OnParametersChanged(object sender, EventArgs e) {
            // Set selected range distribution and factor if it is needed.
            switch (comboRangeDistribution.SelectedIndex) {
                case LogarithmicDistributionIndex:
                    tbFactor.Enabled = true;
                    Colorizer.RangeDistribution = new LogarithmicRangeDistribution() { Factor = GetFactor() };
                    break;
                case LinearDistributionIndex:
                    tbFactor.Enabled = true;
                    Colorizer.RangeDistribution = new ExponentialRangeDistribution() { Factor = GetFactor() };
                    break;
                default:
                    Colorizer.RangeDistribution = new LinearRangeDistribution();
                    tbFactor.Enabled = false;
                    break;
            }
        }

        private double GetFactor() {
            return (String.IsNullOrEmpty(tbFactor.Text)) ? 0 : Convert.ToDouble(tbFactor.Text);
        }
    }
}
